feat: allow hiding offline and/or non-direct nodes from list and map

lkosson opened the pull request:
This change allows filtering out offline (as per existing `NodeInfo.isOnline` logic) and non-directly-reachable (i.e. those with `NodeInfo.hopsAway` greater than zero) nodes from the node list and map.

<img src="https://github.com/user-attachments/assets/5b9192fd-8a3e-44c7-801c-c403a03a2008" width="300"/>

jamesarich commented on app/src/main/java/com/geeksville/mesh/model/UIState.kt:
```diff
@@ -285,21 +290,37 @@ class UIViewModel @Inject constructor(
+        val profile = flows[6] as DeviceProfile
         NodesUiState(
-            sort = sort,
-            filter = filter,
-            includeUnknown = includeUnknown,
+            sort = flows[1] as NodeSortOption,
+            filter = flows[0] as String,
+            includeUnknown = flows[2] as Boolean,
+            onlyOnline = flows[3] as Boolean,
+            onlyDirect = flows[4] as Boolean,
             gpsFormat = profile.config.display.gpsFormat.number,
             distanceUnits = profile.config.display.units.number,
             tempInFahrenheit = profile.moduleConfig.telemetry.environmentDisplayFahrenheit,
-            showDetails = showDetails,
+            showDetails = flows[5] as Boolean,
```
I'm kinda with Copilot on this one - I think we're losing some context by condensing these.

lkosson replied:
`combine` has only overloads for up to 5 parameters, and so does destructuring operator for `Array`, so neither `sort, filter, includeUnknown, onlyOnline, onlyDirect, showDetails, profile ->` nor `(sort, filter, includeUnknown, onlyOnline, onlyDirect, showDetails, profile) ->` works. Any suggestions on how to achieve it? I haven't really used Kotlin before.

jamesarich reviewed (changes requested):
Looks like we need to update `NodeInfoDaoTest` to inclcude the new fields (and potentially some additional tests for the new fields).

jamesarich commented:
> I'm having some problems running test suite locally. What is the correct way to invoke them? `gradlew test` fails with compilation errors (mostly unresolved references in `mesh_service_example`), even on clean checkout from repo tip.

CI is using the command `./gradlew connectedFdroidDebugAndroidTest` which requires a running emulator. You can also run this test from within Android Studio:
![image](https://github.com/user-attachments/assets/ccbd58d1-ba1b-4d1c-a7dd-a04ffd22a7c3)

or

![image](https://github.com/user-attachments/assets/b2066f7d-83d1-4ddf-a447-08f5c5a462d3)

lkosson commented:
I have added new tests and updated `getNodes`. There seems to be a problem with existing `testIncludeUnknownIsFalse` - `Node.isUnknownUser` checks `hwModel` for `HardwareModel.UNSET`, while acutal implementation of `NodeInfoDao` uses `short_name` to determine whether to include a node when `includeUnknown` is set.

Is it a bug, or I am missing something? If it is a bug, should I fix it in this PR, or file a new issue? If I should fix it, which behavior is correct - using `hwModel` or `short_name` for determining `isUnknownUser`?

jamesarich commented:
> Is it a bug, or I am missing something? If it is a bug, should I fix it in this PR, or file a new issue? If I should fix it, which behavior is correct - using `hwModel` or `short_name` for determining `isUnknownUser`?

Good catch - if you want to add that to this PR 👍 , that'll save some legwork with making a new issue.

lkosson commented:
Is there any piece of documentation or explanation when to use User's or Node's shortName?

And - kind of related - what is the difference between `NodeInfo` and `Node`?

jamesarich commented on app/src/main/java/com/geeksville/mesh/database/entity/NodeEntity.kt:
```diff
@@ -45,6 +46,7 @@ data class NodeWithRelations(
             num = num,
             metadata = metadata?.proto,
             user = user,
+            shortName = shortName,
```
we can just use `user.shortName here` no need for an additional field

jamesarich commented on app/src/main/java/com/geeksville/mesh/model/Node.kt:
```diff
@@ -35,6 +35,7 @@ data class Node(
     val num: Int,
     val metadata: MeshProtos.DeviceMetadata? = null,
     val user: MeshProtos.User = MeshProtos.User.getDefaultInstance(),
+    var shortName: String? = null,
```
again we should be able to use `user.shortName` here

jamesarich replied:
:thinking: i'm unfamiliar with this filter/test as well.. why don't we roll back the changes made to address this change bug, and we can break it off into it's own issue and give it a little more thought. I don't want it to hold back the rest of your PR.

lkosson replied:
I have reverted changes to `isUnknownUser` and change mock up data to better fit existing logic.
